src/inference/graph_inference.cpp

```diff
@@ -921,51 +924,46 @@ class GraphNodeAssignment : public Nice2Assignment {
         }
       }
       for (size_t j = 0; j < factors_candidates.size(); ++j) {
-        Factor current_candidate = factors_candidates[j];
-        std::set<int> distinct_labels;
+        std::unordered_map<int,int> distinct_labels;
         std::vector<int> candidate_inf_labels;
-        candidate_inf_labels.reserve(current_candidate.size());
-        std::unordered_set<std::vector<int>> permutations;
-        for (auto label = current_candidate.begin(); label != current_candidate.end(); ++label) {
-          if (giv_labels.count(*label) == 0) {
+        candidate_inf_labels.reserve(factors_candidates[j].size());
+        for (auto label = factors_candidates[j].begin(); label != factors_candidates[j].end(); ++label) {
+          if (giv_labels.count(*label) + distinct_labels[*label] <= factors_candidates[j].count(*label)) {
             candidate_inf_labels.push_back(*label);
-            distinct_labels.insert(*label);
+            if (distinct_labels.count(*label) == 0) {
```

Comment:
You don't need to do this if (this makes 2 lookups, you need one). Just do:  distinct_labels[*label] += 1; (or write it as ++).